src/power.jl

```diff
@@ -58,38 +58,105 @@ end
 
 ^(a::Taylor1{TaylorN{T}}, r::Rational) where {T<:NumberNotSeries} = a^(r.num/r.den)
 
-
-
-# power_by_squaring; slightly modified from base/intfuncs.jl
-# Licensed under MIT "Expat"
-for T in (:Taylor1, :HomogeneousPolynomial, :TaylorN)
-    @eval function power_by_squaring(x::$T, p::Integer)
-        if p == 1
-            return copy(x)
-        elseif p == 0
-            return one(x)
-        elseif p == 2
-            return square(x)
-        end
+# in-place form of power_by_squaring
+# this method assumes `y`, `x` and `aux` are of same order
+for T in (:Taylor1, :TaylorN)
+    @eval function power_by_squaring!(y::$T{T}, x::$T{T}, aux::$T{T},
+            p::Integer) where {T<:NumberNotSeries}
         t = trailing_zeros(p) + 1
         p >>= t
+        # aux = x
+        for k in eachindex(aux)
+            identity!(aux, x, k)
+        end
         while (t -= 1) > 0
-            x = square(x)
+            # aux = square(aux)
+            for k in reverse(eachindex(aux))
+                sqr!(aux, k)
+            end
+        end
+        # y = aux
+        for k in eachindex(y)
+            identity!(y, aux, k)
         end
-        y = x
         while p > 0
             t = trailing_zeros(p) + 1
             p >>= t
             while (t -= 1) ≥ 0
-                x = square(x)
+                # aux = square(aux)
+                for k in reverse(eachindex(aux))
+                    sqr!(aux, k)
+                end
             end
-            y *= x
+            # y = y * aux
+            mul!(y, aux)
         end
-        return y
+        return nothing
     end
 end
 
 
+# power_by_squaring; slightly modified from base/intfuncs.jl
+# Licensed under MIT "Expat"
+for T in (:Taylor1, :HomogeneousPolynomial, :TaylorN)
+    @eval begin
+        power_by_squaring(x::$T, p::Integer) = power_by_squaring(x, Val(p))
+        power_by_squaring(x::$T, ::Val{0}) = one(x)
+        power_by_squaring(x::$T, ::Val{1}) = copy(x)
+        power_by_squaring(x::$T, ::Val{2}) = square(x)
+        if $T != HomogeneousPolynomial
+            function power_by_squaring(x::$T, ::Val{P}) where P
+                p = P # copy static parameter `P` into local variable `p`
+                y = zero(x)
+                aux = zero(x)
+                power_by_squaring!(y, x, aux, p)
+                return y
+            end
```

Comment:
I think this `if` is overwritten for `HomogeneousPolynomial` by the method defined in line 118...

Reply:
Incidentally, I think this code in nicer than the one in line 118...